src/WallpaperEngine/Input/CMouseInput.cpp

```diff
@@ -5,10 +5,20 @@ using namespace WallpaperEngine::Input;
 
 CMouseInput::CMouseInput (GLFWwindow* window) : position (), m_mousePosition (), m_window (window) {}
```

Comment:
For the input I feel like a solution similar to the Render Drivers would be a better approach now that not every input method uses GLFW.

Something like:
- CInputDriver as base class to provide access to raw data like absolute mouse position and the update method
- CGLFWInputDriver that uses the current code
- CXWaylandInputDriver that implements the wayland input reading
- CInputContext that provides access to mouse's x&y position calling one of the CInputDriver and doing the glm::mix that is currently done by CMouseInput

This way the constructors for the input drivers can have any dependency they need (like the video driver as both seem to depend on it) and don't need any casting.

Reply:
we can split them, but the "input reading" is closely tied to the render driver and can't be separated from it, meaning the C_Wayland_ (I assume XWayland was a typo) InputDriver would be pretty much empty and just relaying data from the render driver

Reply:
split. lmk if any further issues arise.

Reply:
> we can split them, but the "input reading" is closely tied to the render driver and can't be separated from it, meaning the C Wayland (I assume XWayland was a typo) InputDriver would be pretty much empty and just relaying data from the render driver

I still feel like that's better than having x11 and wayland specific code in the same class.

> Also please note you cannot get the absolute mouse position in wayland by design, only relative to the layerSurface and only when it's in focus

Then having input separated in different classes based on the driver is going to be a better solution. When running in window mode the mouse position is relative to the window, but while running in background mode, the mouse position is calculated from the absolute mouse position and the current viewport being rendered (on X11 all screens are drawn in the same texture and copied to the screen, so this approach works) and I guess that on Wayland it'll be based on the background surface's, so the viewport information won't be of use for calculations...

This is currently done on WallpaperEngine::Render::CScene::updateMouse but I feel like moving how these calculations are done to the MouseInput would be a better approach to properly handle wayland-specific behaviour. This might need further changes into the Output drivers to return all the required viewport information so the MouseInput knows exactly what surface it needs to calculate positions for as just the position and size of the viewport will not be enough for Wayland...

Reply:
we don't even have the position of the surface. We could _guess_ it by querying xdg_output but that's be a _guess_ and IMHO an unneeded one. So far mouse works, but I haven't gotten multi-monitor to work properly yet so I can't test on that.

I mean, technically it _does run_ with two monitors, but one is black and the other is like 2FPS for... reasons xD